### DISCUSSION TITLE
[C128] Compile with two different versions of the Runtime library modules possible?

### DISCUSSION THREAD

**Discussion by RetrofanC** [cc65/cc65#1932] [Q&A] [answered]
This might go a bit too far and deep but if somebody could give some direction for me to explore further that would be appreciated. 

I am wondering where and how the cc65 toolchain decides to replace certain code of the c source during compilation with standard available library modules like; add.s , pushwsp.s, staxspi.s 
(Is this done by the ca65.exe source or some of the other parsers within the cc65 toolchain perhaps?) 

Reason is that for my memory switching project, especially with more complex functions called. 
It seems I need two versions for some of these library modules. 

One standard version, for the main memory area and an alternative version; add_common.s, pushwsp_common.s, staxspi_common.s for the functions running in RAM#1, RAM#2 & RAM#3. 

So the thinking is that there will be added some ‘extra checks’ that when say the #pragma code-name (push, “RAMMEMx”) is used in the c source. The compiler should not use the standard “add.s” but will use the “add_common.s” instead etc.

**Comment by mrdudz** [comment, marked as answer]
> I am wondering where and how the cc65 toolchain decides to replace certain code of the c source during compilation with standard available library modules like; add.s , pushwsp.s, staxspi.s

The codegenerator in cc65 does that (see codegen.c)

> One standard version, for the main memory area and an alternative version; add_common.s, pushwsp_common.s, staxspi_common.s for the functions running in RAM#1, RAM#2 & RAM#3.

A common problem with banked code - and unfortunately there is no good solution. Right now the only way that works "out of the box" without further trickery is to place all those functions (ie the entire .code segment) in common (non banked) memory.

To achive what you really want, ie have identical copies of that code in different banks, you will need some advanced hackery. For example compile the extra banks into binaries seperately first, and then glue everything together in an extra step. That will usually involve a script or two that parses eg the main binary file, and creates "import" library headers with absolute addresses in it for the other banks.

**Reply by RetrofanC** [reply]
Ahh most helpful! (-8 

I had a look at the codegen.c and the others in that directory which indeed are the ones I was looking for.
Perhaps during the Christmas break I can look into that a bit more deeper and see how I can adopt that. 

The current structure I have which is working good for now is to have a C128.LIB (regular) and a C128_COMMON.LIB. 
The C128_COMMON.LIB does contain all modules which have to run in common memory. 

When I can get to a situation where the majority of the (semi-complex) functions (to be called from foreign banks) can be served from the C128_COMMON.LIB we are in good waters. Just now manual finding out how many of them are needed, which are working without problems and what size the common memory needs to accommodate them. With the current 8k I am doing rather well. But even if I have to make the common a 16k size I still have 3x 48k (RAM#1,#2,#3) to win by having full seamless banking for my applications. 

When the above approach gives too much limitations I will take up your suggestion for sure to look into scripting and pre-loading 'Library Modules Blocks" in every RAM#n at the same location. 

To be continued but sofar I have new area's and idea's to work on thanks!
